modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Fields.tsx

```diff
@@ -306,15 +312,165 @@ const SaveFDSFieldsModalContent = ({
 	);
 };
 
+interface EditFDSFieldModalContentProps {
+	closeModal: Function;
+	fdsField: FDSField;
+	namespace: string;
+	onSave: Function;
+}
+
+const EditFDSFieldModalContent = ({
+	closeModal,
+	fdsField,
+	namespace,
+	onSave,
+}: EditFDSFieldModalContentProps) => {
+	const [selectedFDSFieldRenderer, setSelectedFDSFieldRenderer] = useState(
+		fdsField.renderer ?? 'default'
+	);
+	const [fdsFieldSortable, setFSDFieldSortable] = useState<boolean>(
+		fdsField.sortable ?? true
+	);
+
+	const fdsFieldLabelRef = useRef<HTMLInputElement>(null);
```

Comment:
I really don't think this is a good pattern to follow... It feels very anti-react to use refs to grab our values on submit. I'm happy to be wrong, but I really haven't seen other people do this or it being provided as a prescribed way to do things in the react docs. [They recommend either controlled components or form](https://legacy.reactjs.org/docs/uncontrolled-components.html)

If we don't want the input to be controlled via state, we can use a `<form>`(or ClayForm) instead and then just grab the data from that onSubmit and put our submit button within that element.

Reply:
The reason for not using state is performance. An `onChange` updated states on each change, causing refreshes of the whole component. As this is a free input text field, if you press and hold a key, it'll result in many many refreshes per second. Sure, we can reduce the issue with throttling / debouncing, but it's better to completely remove the issue by using a ref. I agree that the state approach is cleaner ant more react-y, that's why I am using in on the select and checkbox elements in this PR.

The form approach from the link above is basically what we are doing in this PR. They are using refs as well. We are just not using the `form` element and `submit` action, rather constructing a fetch request manually on click.

Reply:
> The reason for not using state here is performance

Is this actually an issue though? Because if "controlled" form elements were a performance hit, I don't really think that they would be the recommended way of doing things, which the react docs seem to indicate. And yes the component "re-renders" on every keystroke, but only the "input" element should be refreshing in the DOM because the VDOM should recognize that nothing else is changed except the input of the element.

My concern is that future maintenance of code where we use "ref" instead of "state" will be much harder to reason through and maintain because by nature, "ref" is a mutable variable and "state" is not. I would prefer we only reach for "ref" when we need to hold a value that is unrelated to the UI or render lifecycle. In your case here, it is simple, but as soon as we need to add another input or any sort of validation or response to that value, we end up having to refactor to state.

This might be a larger discussion as a whole that we bring to the infra team, as we likely need to make a "standard" of when to use "ref vs state vs context vs props", otherwise each component starts to look slightly different depending on who is creating it.

Reply:
> And yes the component "re-renders" on every keystroke, but only the "input" element should be refreshing in the DOM because the VDOM should recognize that nothing else is changed except the input of the element

I'm afraid React is not that smart. Evarything in the scope of state variable re-renders:

https://github.com/liferay-frontend/liferay-portal/assets/5435511/15c5aa15-7d91-468a-96d1-7fe706b0388d

Reply:
I am actually not sure what yellow border of devtools highlight means, it activates on large frequency of input :thinking: But the scope of effect for state is the whole component:

https://github.com/liferay-frontend/liferay-portal/assets/5435511/2a16a77a-29cc-49b4-a0e6-1c063789f8ab

Reply:
> My concern is that future maintenance of code where we use "ref" instead of "state" will be much harder to reason through and maintain because by nature, "ref" is a mutable variable and "state" is not. I would prefer we only reach for "ref" when we need to hold a value that is unrelated to the UI or render lifecycle. In your case here, it is simple, but as soon as we need to add another input or any sort of validation or response to that value, we end up having to refactor to state.

I agree, guidelines would be nice. And I also agree that a set of states should be prefered as default way to implement UI. But, I think there are valid cases for alternatives, like ref, but also reducer, or context. I'd say that free input is one of those cases.

Reply:
> Everything in the scope of state variable re-renders

Ah yeah, sorry I think I mispoke here a bit. Re-renders happen, meaning that react runs the components lifecycle again. What I actually meant was "re-painting" in the browser. So when react is reconciling the dom vs vdom, only the input with the change of value repaints.


> I am actually not sure what yellow border of devtools highlight means, it activates on large frequency of input 🤔 But the scope of effect for state is the whole component:

I wonder if this happens because of FocusScope within clay... Possible that a parent "context" is changing state and causing the whole tree to execute again



-----------

On the topic of avoiding re-renders, I vaguely remember this conversation come up on twitter with the react core team and they indicated that it's likely not necessary to over optimize and avoid them, but only seek to optimize in cases where there is a noticeable performance issue. It was possible this same up not because of refs but because of useMemo and useCallback, where you could easily apply those everywhere but the performance benefit actually doesn't outweigh the readability impact. So they recommended to only optimize re-renders when an issue is seen.

I can't seem to find these conversations though... We also had a few ourselves when Greg was around but now I can't seem to find them at all. 🤷 



FYI: In the end, I personally don't think I would use a ref like this, but its also not a blocker for merge and if you think it's best, than go for it. I've actually been thinking a lot about forms like this and how we can possibly simplify the logic behind them. We can always refactor ref/state/whatever in the future if we find a better way.